File: hnaj-be/app/Http/Controllers/Api/Auth/EmailVerificationController.php
<?php

namespace App\Http\Controllers\Api\Auth;

use App\Actions\Auth\ResendEmailVerification;
use App\Actions\Auth\VerifyEmail;
use App\Http\Controllers\Controller;
use App\Http\Requests\Auth\ResendVerificationRequest;
use App\Http\Requests\Auth\VerifyEmailRequest;
use App\Http\Resources\UserResource;
use App\Http\Responses\ApiResponse;
use Illuminate\Http\JsonResponse;

class EmailVerificationController extends Controller
{
    public function verify(VerifyEmailRequest $request, VerifyEmail $verifyEmail): JsonResponse
    {
        $user = $verifyEmail->handle($request->token());

        return ApiResponse::success(
            data: ['user' => new UserResource($user)],
            message: 'Email verified successfully. You can sign in now.',
        );
    }

    public function resend(ResendVerificationRequest $request, ResendEmailVerification $resend): JsonResponse
    {
        $resend->handle($request->email());

        // Thông điệp trung lập để không tiết lộ email nào đã đăng ký.
        return ApiResponse::success(
            message: 'If the email address needs verification, a new link has been sent.',
        );
    }
}

File: hnaj-be-java/src/main/java/com/hnaj/auth/controller/EmailVerificationController.java
package com.hnaj.auth.controller;

import com.hnaj.api.common.ApiResponse;
import com.hnaj.auth.ResendEmailVerificationService;
import com.hnaj.auth.UserView;
import com.hnaj.auth.VerifyEmailService;
import com.hnaj.auth.validation.AuthInput;
import java.util.Map;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

@RestController
public class EmailVerificationController {
    private final VerifyEmailService verify;
    private final ResendEmailVerificationService resend;

    public EmailVerificationController(
            VerifyEmailService verify,
            ResendEmailVerificationService resend) {
        this.verify = verify;
        this.resend = resend;
    }

    @PostMapping("/api/auth/email/verify")
    public ResponseEntity<Map<String, Object>> verify(@RequestBody(required = false) Map<String, Object> body) {
        AuthInput input = new AuthInput(body);
        String token = input.required("token", 255, false, false);
        input.validate();

        UserView user = verify.verify(token);

        return ApiResponse.success(Map.of("user", user),
                "Email verified successfully. You can sign in now.", null);
    }

    @PostMapping("/api/auth/email/resend")
    public ResponseEntity<Map<String, Object>> resend(@RequestBody(required = false) Map<String, Object> body) {
        AuthInput input = new AuthInput(body);
        String email = input.email();
        input.validate();

        resend.resend(email);

        return ApiResponse.success(null,
                "If the email address needs verification, a new link has been sent.", null);
    }
}